File: akka-http-marshallers-java/akka-http-jackson/src/main/scala/akka/http/javadsl/marshallers/jackson/Jackson.scala
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.http.javadsl.marshallers.jackson

import com.fasterxml.jackson.databind.{ MapperFeature, ObjectMapper }
import scala.reflect.ClassTag
import akka.http.scaladsl.marshalling
import akka.http.scaladsl.unmarshalling
import akka.http.scaladsl.model.MediaTypes._
import akka.http.javadsl.server.{ Unmarshaller, Marshaller }
import akka.http.impl.server.{ UnmarshallerImpl, MarshallerImpl }

object Jackson {
  private val objectMapper: ObjectMapper = new ObjectMapper().enable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY)
  def json[T <: AnyRef]: Marshaller[T] = jsonMarshaller(objectMapper).asInstanceOf[Marshaller[T]]
  def json[T <: AnyRef](objectMapper: ObjectMapper): Marshaller[T] = jsonMarshaller(objectMapper).asInstanceOf[Marshaller[T]]
  def jsonAs[T](clazz: Class[T]): Unmarshaller[T] = jsonAs(objectMapper, clazz)
  def jsonAs[T](objectMapper: ObjectMapper, clazz: Class[T]): Unmarshaller[T] =
    UnmarshallerImpl[T] {
      unmarshalling.Unmarshaller.messageUnmarshallerFromEntityUnmarshaller { // isn't implicitly inferred for unknown reasons
        unmarshalling.Unmarshaller.stringUnmarshaller
          .forContentTypes(`application/json`)
          .map { jsonString ⇒
            val reader = objectMapper.reader(clazz)
            clazz.cast(reader.readValue(jsonString))
          }
      }
    }(ClassTag(clazz))

  private def jsonMarshaller(objectMapper: ObjectMapper): Marshaller[AnyRef] =
    MarshallerImpl[AnyRef] { implicit ec ⇒
      marshalling.Marshaller.StringMarshaller.wrap(`application/json`) { (value: AnyRef) ⇒
        val writer = objectMapper.writer()
        writer.writeValueAsString(value)
      }
    }
}

File: akka-http-marshallers-java/akka-http-jackson/src/main/java/akka/http/javadsl/marshallers/jackson/Jackson.java
/*
 * Copyright (C) 2009-2016 Typesafe Inc. <http://www.typesafe.com>
 */
package akka.http.javadsl.marshallers.jackson;

import java.io.IOException;

import akka.http.javadsl.model.HttpEntity;
import akka.http.javadsl.model.MediaTypes;
import akka.http.javadsl.model.RequestEntity;
import akka.http.javadsl.server.Marshaller;
import akka.http.javadsl.server.Unmarshaller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

public class Jackson {
	private static final ObjectMapper defaultObjectMapper = 
			new ObjectMapper().enable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY);
	
	public static <T> Marshaller<T, RequestEntity> marshaller() {
		return marshaller(defaultObjectMapper);
	}
	
	public static <T> Marshaller<T, RequestEntity> marshaller(ObjectMapper mapper) {
		return Marshaller.wrapEntity(
            u -> toJSON(mapper, u),
            Marshaller.stringToEntity(), 
            MediaTypes.APPLICATION_JSON 
        );
	}
	
	public static <T> Unmarshaller<HttpEntity,T> unmarshaller(Class<T> expectedType) {
		return unmarshaller(defaultObjectMapper, expectedType);
	}
	
	public static <T> Unmarshaller<HttpEntity,T> unmarshaller(ObjectMapper mapper, Class<T> expectedType) {
        return Unmarshaller.forMediaType(MediaTypes.APPLICATION_JSON, Unmarshaller.entityToString())
	        .map(s -> fromJSON(mapper, s, expectedType));		
	}
	
	private static String toJSON(ObjectMapper mapper, Object object) {
		try {
			return mapper.writeValueAsString(object);
		} catch (JsonProcessingException e) {
			throw new IllegalArgumentException("Cannot marshal to JSON: " + object, e);
		}
	}
	
	private static <T> T fromJSON(ObjectMapper mapper, String json, Class<T> expectedType) {
		try {
			return mapper.reader(expectedType).readValue(json);
		} catch (IOException e) {
			throw new IllegalArgumentException("Cannot unmarshal JSON as " + expectedType.getSimpleName(), e);
		}
	}
}
